File: research/experiments/sweep/runs/opt-medium-csharp-sonnet45/Optimization/FiniteDiff.cs
namespace Optimization;

/// <summary>
/// Finite difference gradient approximation methods.
/// </summary>
public static class FiniteDiff
{
    private const double MachineEpsilon = 2.22e-16;

    /// <summary>
    /// Approximate gradient using forward differences.
    /// Step size: h = √ε × max(|xᵢ|, 1)
    /// </summary>
    public static double[] ForwardDiffGradient(Func<double[], double> f, double[] x)
    {
        int n = x.Length;
        double[] gradient = new double[n];
        double fx = f(x);

        for (int i = 0; i < n; i++)
        {
            double h = Math.Sqrt(MachineEpsilon) * Math.Max(Math.Abs(x[i]), 1.0);
            double[] xPlusH = VecOps.Clone(x);
            xPlusH[i] += h;

            double fPlusH = f(xPlusH);
            gradient[i] = (fPlusH - fx) / h;
        }

        return gradient;
    }

    /// <summary>
    /// Approximate gradient using central differences.
    /// Step size: h = ∛ε × max(|xᵢ|, 1)
    /// </summary>
    public static double[] CentralDiffGradient(Func<double[], double> f, double[] x)
    {
        int n = x.Length;
        double[] gradient = new double[n];

        for (int i = 0; i < n; i++)
        {
            double h = Math.Pow(MachineEpsilon, 1.0 / 3.0) * Math.Max(Math.Abs(x[i]), 1.0);
            double[] xPlusH = VecOps.Clone(x);
            double[] xMinusH = VecOps.Clone(x);
            xPlusH[i] += h;
            xMinusH[i] -= h;

            double fPlusH = f(xPlusH);
            double fMinusH = f(xMinusH);
            gradient[i] = (fPlusH - fMinusH) / (2.0 * h);
        }

        return gradient;
    }

    /// <summary>
    /// Factory function that returns a gradient function using the specified method.
    /// </summary>
    public static Func<double[], double[]> MakeGradient(
        Func<double[], double> f,
        string method = "forward")
    {
        return method switch
        {
            "central" => x => CentralDiffGradient(f, x),
            "forward" => x => ForwardDiffGradient(f, x),
            _ => throw new ArgumentException($"Unknown method: {method}")
        };
    }
}

File: research/experiments/sweep/runs/opt-large-python-sonnet45/src/finite_diff.py
"""Numerical gradient approximation via finite differences."""

import sys
import math
from typing import Callable, List


EPS = sys.float_info.epsilon


def forward_diff_gradient(f: Callable[[List[float]], float], x: List[float]) -> List[float]:
    """Approximate gradient using forward differences."""
    n = len(x)
    grad = [0.0] * n
    fx = f(x)

    for i in range(n):
        h = math.sqrt(EPS) * max(abs(x[i]), 1.0)
        x_perturbed = x[:]
        x_perturbed[i] += h
        grad[i] = (f(x_perturbed) - fx) / h

    return grad


def central_diff_gradient(f: Callable[[List[float]], float], x: List[float]) -> List[float]:
    """Approximate gradient using central differences."""
    n = len(x)
    grad = [0.0] * n

    for i in range(n):
        h = EPS ** (1.0 / 3.0) * max(abs(x[i]), 1.0)
        x_plus = x[:]
        x_plus[i] += h
        x_minus = x[:]
        x_minus[i] -= h
        grad[i] = (f(x_plus) - f(x_minus)) / (2.0 * h)

    return grad


def make_gradient(
    f: Callable[[List[float]], float],
    method: str = "forward"
) -> Callable[[List[float]], List[float]]:
    """Factory function that returns a gradient function."""
    if method == "forward":
        return lambda x: forward_diff_gradient(f, x)
    elif method == "central":
        return lambda x: central_diff_gradient(f, x)
    else:
        raise ValueError(f"Unknown method: {method}")
